File: Thpolish/py/depth.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sys
import os
import argparse
import pysam
import numpy as np
import time
from multiprocessing import Pool

# 直方图统计上限，用于计算中位数
HIST_MAX_DEPTH = 1000 

CLIP_MIN = -1.0
CLIP_MAX = 3.0

def worker_compute_chrom(args):
    bam_path, chrom, length = args
    
    # 1. 尝试打开 BAM 文件
    try:
        samfile = pysam.AlignmentFile(bam_path, "rb")
    except Exception as e:
        return chrom, None, None, str(e)

    # 2. 检查索引是否存在
    if not samfile.has_index():
        samfile.close()
        return chrom, None, None, "No Index (.bai/.csi) found"

    try:
        # 3. 差分数组计算深度
        diff_array = np.zeros(length + 1, dtype=np.int32)
        read_count = 0
        
        for read in samfile.fetch(chrom):
            if read.is_unmapped or read.is_secondary or read.is_supplementary:
                continue
            
            read_count += 1
            start = read.reference_start
            end = read.reference_end if read.reference_end else (start + read.query_length)
            
            if start < 0: start = 0
            if end > length: end = length
            
            if start < end:
                diff_array[start] += 1
                diff_array[end] -= 1
        
        samfile.close()

        # 4. 如果没有 Reads，直接返回 None，节省内存
        if read_count == 0:
            return chrom, None, None, None

        # 5. 还原深度 (CumSum)
        depth_array = np.cumsum(diff_array, dtype=np.int32)
        final_depth = depth_array[:-1] # 去掉辅助位

        # 6. 计算局部直方图 (用于后续合并计算全局中位数)
        clipped_depth = np.clip(final_depth, 0, HIST_MAX_DEPTH)
        hist = np.bincount(clipped_depth, minlength=HIST_MAX_DEPTH+1)

        # 处理直方图溢出
        if len(hist) > HIST_MAX_DEPTH + 1:
            overflow = np.sum(hist[HIST_MAX_DEPTH+1:])
            hist = hist[:HIST_MAX_DEPTH+1]
            hist[-1] += overflow
            
        return chrom, final_depth, hist, None

    except Exception as e:
        return chrom, None, None, str(e)

def calc_global_stats(global_hist):
    """根据全局直方图计算统计量"""
    total_bases = np.sum(global_hist)
    if total_bases == 0:
        return 0.0, 0.0

    # 计算均值
    depths = np.arange(len(global_hist))
    total_coverage = np.sum(depths * global_hist)
    global_mean = total_coverage / total_bases

    # 计算中位数
    cumsum = np.cumsum(global_hist)
    median_idx = np.searchsorted(cumsum, total_bases / 2)
    global_median = float(median_idx)

    return global_mean, global_median

def main():
    parser = argparse.ArgumentParser(description="Calculate Zero-Centered Normalized Depth")
    parser.add_argument("-b", "--bam", required=True, help="Input BAM file")
    parser.add_argument("-o", "--out", required=True, help="Output .npz file")
    parser.add_argument("-t", "--threads", type=int, default=4)
    parser.add_argument("-l", "--min_len", type=int, default=0)
    parser.add_argument("--chroms", nargs="+", help="Specific chromosomes (optional)")
    
    args = parser.parse_args()
    t0 = time.time()

    # 文件检查
    if not os.path.exists(args.bam):
        sys.exit(f"File not found: {args.bam}")
    
    # 简单的索引文件检查警告
    if not os.path.exists(args.bam + ".bai") and not os.path.exists(args.bam + ".csi"):
        print(f"[Warning] BAM index not found nearby. Fetch might fail.")

    print(f"[Init] Reading header...")
    with pysam.AlignmentFile(args.bam, "rb") as f:
        all_refs = f.references
        all_lens = f.lengths
        ref_map = dict(zip(all_refs, all_lens))

    target_chroms = args.chroms if args.chroms else all_refs
    tasks = []
    for c in target_chroms:
        if c in ref_map and ref_map[c] >= args.min_len:
            tasks.append((args.bam, c, ref_map[c]))

    print(f"[Run] Step 1: Calculating raw depth (Threads: {args.threads})...")
    
    # 暂存原始数据: {chrom: int32_array}
    raw_data_store = {}
    
    # 全局直方图
    global_hist = np.zeros(HIST_MAX_DEPTH + 1, dtype=np.int64)
    total_len = 0
        
    with Pool(processes=args.threads) as pool:
        for chrom, depth_arr, local_hist, err in pool.imap_unordered(worker_compute_chrom, tasks):
            if err:
                print(f"  > [Error] {chrom}: {err}")
                continue
            
            if depth_arr is not None:
                raw_data_store[chrom] = depth_arr
                
                l = min(len(local_hist), len(global_hist))
                global_hist[:l] += local_hist[:l]
                total_len += len(depth_arr)
                
                print(f"  > [Loaded] {chrom:<10} (Mean Raw: {np.mean(depth_arr):.1f}x)")

    # --- Step 2: 全局统计 ---
    print("\n[Stats] Calculating Global Median...")
    g_mean, g_median = calc_global_stats(global_hist)
    
    print(f" Total Bases   : {total_len}")
    print(f" Global Mean   : {g_mean:.4f}")
    print(f" Global Median : {g_median:.4f} (Baseline)")

    if g_median < 1.0:
        print("[Warning] Global median is < 1.0. Defaulting to 1.0 to avoid error.")
        g_median = 1.0

    # --- Step 3: 归一化 (Zero-Centered) ---
    print(f"\n[Run] Step 2: Normalizing (Zero-Centered) & Converting to float16...")
    print(f"  > Formula: (Raw / Median) - 1.0")
    print(f"  > Clip Range: [{CLIP_MIN}, {CLIP_MAX}]")

    chrom_keys = list(raw_data_store.keys())
    for chrom in chrom_keys:
        raw_arr = raw_data_store[chrom]
        
        norm_arr = (raw_arr / g_median) - 1.0
        
        norm_arr = np.clip(norm_arr, CLIP_MIN, CLIP_MAX)
        
        raw_data_store[chrom] = norm_arr.astype(np.float16)
        
        del raw_arr
        
    print(f"  > Normalization done.")

    # --- Step 4: 保存 ---
    meta_info = {
        "global_median": g_median,
        "global_mean": g_mean,
        "normalization": "zero-centered",
        "clip_min": CLIP_MIN,
        "clip_max": CLIP_MAX
    }
    
    print(f"[Save] Writing compressed data to {args.out} ...")
    np.savez_compressed(args.out, **raw_data_store, **meta_info)
    
    print(f"[Finished] Total time: {time.time() - t0:.2f}s")

if __name__ == "__main__":
    main()

File: Thpolish/src/depth.rs
use anyhow::{Context, Result};
use log::info;
use rayon::prelude::*;
use rust_htslib::bam::{self, ext::BamRecordExtensions, Read};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::Path;

use crate::option::DepthArgs;

const HIST_MAX_DEPTH: usize = 1000;
const CLIP_MIN: f32 = -1.0;
const CLIP_MAX: f32 = 3.0;

#[derive(Serialize, Deserialize, Debug)]
pub struct DepthMetadata {
    pub global_median: f32,
    pub global_mean: f32,
    pub normalization: String,
    pub clip_min: f32,
    pub clip_max: f32,
}

#[derive(Serialize, Deserialize)]
pub struct DepthArchive {
    pub metadata: DepthMetadata,
    pub depths: HashMap<String, Vec<f32>>,
}

pub fn run_depth(args: DepthArgs) -> Result<()> {
    info!("Depth computation initialized. Target BAM: {}", args.bam);

    let bam_reader = bam::IndexedReader::from_path(&args.bam)
        .context("Failed to open BAM or missing index (.bai/.csi)")?;
    let header = bam_reader.header().to_owned();
    
    let mut target_chroms = Vec::new();
    let ref_names = header.target_names();
    
    for (tid, name_bytes) in ref_names.into_iter().enumerate() {
        let name = String::from_utf8_lossy(name_bytes).to_string();
        let len = header.target_len(tid as u32).unwrap_or(0) as usize;
        
        if len < args.min_len { continue; }
        if let Some(ref allowed) = args.chroms {
            if !allowed.contains(&name) { continue; }
        }
        target_chroms.push((tid as u32, name, len));
    }

    if target_chroms.is_empty() {
        anyhow::bail!("No chromosomes matched the criteria.");
    }

    info!("Calculating raw depth (Threads: {})...", args.thread);
    
    if args.thread > 0 {
        rayon::ThreadPoolBuilder::new()
            .num_threads(args.thread.min(target_chroms.len()))
            .build_global().ok();
    }

    let results: Vec<_> = target_chroms.into_par_iter().filter_map(|(tid, name, len)| {
        match compute_chrom(&args.bam, tid, len) {
            Ok(Some((depth_arr, local_hist))) => {
                info!("  > [Loaded] {:<10} (Length: {})", name, len);
                Some((name, depth_arr, local_hist))
            },
            Ok(None) => None, 
            Err(e) => {
                panic!("  > [Error] {}: {}", name, e)
            }
        }
    }).collect();

    let mut global_hist = [0u64; HIST_MAX_DEPTH + 1];
    let mut total_len = 0usize;
    let mut raw_data_store = HashMap::new();

    for (name, depth_arr, local_hist) in results {
        for i in 0..=HIST_MAX_DEPTH {
            global_hist[i] += local_hist[i];
        }
        total_len += depth_arr.len();
        raw_data_store.insert(name, depth_arr);
    }

    let (g_mean, mut g_median) = calc_global_stats(&global_hist, total_len);
    info!("\n[Stats] Total Bases   : {}", total_len);
    info!("[Stats] Global Mean   : {:.4}", g_mean);
    info!("[Stats] Global Median : {:.4} (Baseline)", g_median);

    if g_median < 1.0 {
        info!("[Warning] Global median is < 1.0. Defaulting to 1.0.");
        g_median = 1.0;
    }

    info!("\nNormalizing and clamping data...");
    let mut final_depths = HashMap::with_capacity(raw_data_store.len());
    
    for (chrom, raw_arr) in raw_data_store {
        let norm_arr: Vec<f32> = raw_arr.into_par_iter()
            .map(|val| {
                let norm = (val as f32 / g_median) - 1.0;
                norm.clamp(CLIP_MIN, CLIP_MAX)
            })
            .collect();
        
        final_depths.insert(chrom, norm_arr);
    }

    info!("Writing compressed archive to {} ...", args.out);
    let archive = DepthArchive {
        metadata: DepthMetadata {
            global_median: g_median,
            global_mean: g_mean,
            normalization: "zero-centered".to_string(),
            clip_min: CLIP_MIN,
            clip_max: CLIP_MAX,
        },
        depths: final_depths,
    };

    let out_stream = std::io::stdout().lock();
    let mut encoder = zstd::Encoder::new(out_stream, 3)?; 
    bincode::serialize_into(&mut encoder, &archive)
        .context("Failed to serialize depth archive")?;
    let _ = encoder.finish()?;

    info!("Depth normalization complete.");
    Ok(())
}

fn compute_chrom(bam_path: &str, tid: u32, len: usize) -> Result<Option<(Vec<i32>, [u64; HIST_MAX_DEPTH + 1])>> {
    let mut reader = bam::IndexedReader::from_path(Path::new(bam_path))?;
    reader.fetch(tid)?;

    let mut diff_array = vec![0i32; len + 1];
    let mut read_count = 0;
    let mut record = bam::Record::new();

    while let Some(Ok(())) = reader.read(&mut record) {
        if record.is_unmapped() || record.is_secondary() || record.is_supplementary() {
            continue;
        }
        read_count += 1;

        // let start = record.pos() as usize;
        // let end_pos = record.cigar().end_pos();
        // let end = (end_pos as usize).min(len);

        let start = record.reference_start() as usize;
        let end = record.reference_end() as usize;

        if start < end {
            diff_array[start] += 1;
            diff_array[end] -= 1;
        }
    }

    if read_count == 0 {
        return Ok(None);
    }

    let mut hist = [0u64; HIST_MAX_DEPTH + 1];
    let mut current_depth = 0i32;
    for i in 0..len {
        current_depth += diff_array[i];
        diff_array[i] = current_depth;
        
        let bin = (current_depth as usize).min(HIST_MAX_DEPTH);
        hist[bin] += 1;
    }
    
    diff_array.pop();

    Ok(Some((diff_array, hist)))
}

fn calc_global_stats(global_hist: &[u64; HIST_MAX_DEPTH + 1], total_bases: usize) -> (f32, f32) {
    if total_bases == 0 { return (0.0, 0.0); }

    let mut total_coverage = 0u64;
    for (depth, &count) in global_hist.iter().enumerate() {
        total_coverage += depth as u64 * count;
    }
    let mean = total_coverage as f64 / total_bases as f64;

    let target = total_bases as u64 / 2;
    let mut cumsum = 0u64;
    let mut median = 0.0;

    for (depth, &count) in global_hist.iter().enumerate() {
        cumsum += count;
        if cumsum >= target {
            median = depth as f32;
            break;
        }
    }

    (mean as f32, median)
}
